src/cmd/root.go

```diff
@@ -16,8 +16,26 @@
 
 package cmd
 
+/*
+#cgo LDFLAGS: -l subid
+#include <shadow/subid.h>
+#include <stdlib.h>
+#include <stdio.h>
+const char *Prog = "storage";
+FILE *shadow_logfd = NULL;
+struct subid_range get_range(struct subid_range *ranges, int i)
+{
+	shadow_logfd = stderr;
+	return ranges[i];
+}
+#if !defined(SUBID_ABI_MAJOR) || (SUBID_ABI_MAJOR < 4)
+# define subid_get_uid_ranges get_subuid_ranges
+# define subid_get_gid_ranges get_subgid_ranges
+#endif
+*/
+import "C"
```

Comment:
Would it be possible to move the CGO part to a different package or something?  It could go to `github.com/containers/toolbox/pkg/utils` (see `src/pkg/utils/utils.go`), or an entirely separate package for CGO bits (not sure what to call it).  Just like [github.com/containers/storage/pkg/idtools](https://github.com/containers/storage/blob/main/pkg/idtools/idtools_supported.go).

The reason I mention this, is that there are a few other corner cases where we can improve our error handling by using CGO.

It came up in the context of the [newly added](https://github.com/containers/toolbox/issues/1066) `--preserve-fds` option.  If we want to check the validity of the value passed to that option, then we would need a C [constructor](https://gcc.gnu.org/onlinedocs/gcc-4.7.0/gcc/Function-Attributes.html) function (ie., `__attribute__((constructor))`) to validate the file descriptors.  See https://github.com/containers/podman/pull/7125 for the discussion around the same check in Podman.

Another use-case is https://github.com/containers/toolbox/issues/908 because I don't see a ready-made Go wrapper for `getpwnam()` or `getpwuid()` that will give the shell field.  However, [github.com/containers/storage/pkg/idtools](https://github.com/containers/storage/blob/main/pkg/idtools/idtools_unix.go) seems to shell out to `getent(1)` so it might not be that urgent.  🙂

So, having all the gory details of the CGO stuff neatly tucked away in a corner will make easier to hack on the Toolbx sources, review the changes, etc..

Reply:
I can explore this, sure. I think that makes sense to at least try.

Reply:
I see that you moved the CGO bits to `src/cmd/cgo_util.go`.  That's much better.  :)

Would it be possible to move it to some other package than `cmd`?  Maybe inside `utils` as `pkg/utils/utils_cgo.go`.  I am saying this because we have a few [other use-cases](https://github.com/containers/toolbox/issues/832) that could benefit from the use of CGO, and @HarryMichal started this trend of keeping `cmd` restricted only to the front-end user-facing code.  The gory CGO stuff isn't quite that.

Reply:
I tried a simple `git mv ...` to move the file, and after some minor tweaks like starting the function with an uppercase `V`, it works.